blockdag/reachability_test.go

```diff
@@ -0,0 +1,446 @@
+	// Create the root node of a new reachability tree
+	root, err = newReachabilityTreeNode(1, 100)
+	if err != nil {
+		t.Fatalf("TestAddTreeChild: newReachabilityTreeNode failed: %s", err)
+	}
+
+	// Add child nodes to root until a reindex occurs (2^6=64 < 100 < 2^7=128)
+	childNodes := make([]*reachabilityTreeNode, 7)
+	for i := 0; i < len(childNodes); i++ {
+		childNodes[i], err = newReachabilityTreeNode(1, math.MaxUint64-1)
+		if err != nil {
+			t.Fatalf("TestAddTreeChild: newReachabilityTreeNode failed: %s", err)
+		}
+		err = root.addTreeChild(childNodes[i])
+		if err != nil {
+			t.Fatalf("TestAddTreeChild: addTreeChild failed: %s", err)
+		}
+	}
+
+	// Expect the last-added child to have an interval of 1 and remaining interval of 0
+	lastChild := childNodes[len(childNodes)-1]
+	lastChildInterval := lastChild.interval.size()
+	if lastChildInterval != 1 {
+		t.Fatalf("TestAddTreeChild: unexpected lastChild interval size: want: 1, got: %d", lastChildInterval)
+	}
+	lastChildRemainingInterval := lastChild.remainingInterval.size()
+	if lastChildRemainingInterval != 0 {
+		t.Fatalf("TestAddTreeChild: unexpected lastChild interval size: want: 0, got: %d", lastChildRemainingInterval)
+	}
+
+	// Expect all nodes to be descendant nodes of root
+	for _, childNode := range childNodes {
+		if !root.interval.isAncestorOf(&childNode.interval) {
+			t.Fatalf("TestAddTreeChild: childNode is not a descendant of root")
+		}
+	}
+}
+
+func TestSplitFraction(t *testing.T) {
+	tests := []struct {
+		interval      *reachabilityInterval
+		fraction      float64
+		expectedLeft  *reachabilityInterval
+		expectedRight *reachabilityInterval
+	}{
+		{
+			interval:      &reachabilityInterval{start: 1, end: 100},
+			fraction:      0.5,
+			expectedLeft:  &reachabilityInterval{start: 1, end: 50},
+			expectedRight: &reachabilityInterval{start: 51, end: 100},
+		},
+		{
+			interval:      &reachabilityInterval{start: 2, end: 100},
+			fraction:      0.5,
+			expectedLeft:  &reachabilityInterval{start: 2, end: 51},
+			expectedRight: &reachabilityInterval{start: 52, end: 100},
+		},
+		{
+			interval:      &reachabilityInterval{start: 1, end: 99},
+			fraction:      0.5,
+			expectedLeft:  &reachabilityInterval{start: 1, end: 50},
+			expectedRight: &reachabilityInterval{start: 51, end: 99},
+		},
+		{
+			interval:      &reachabilityInterval{start: 1, end: 100},
+			fraction:      0.2,
+			expectedLeft:  &reachabilityInterval{start: 1, end: 20},
+			expectedRight: &reachabilityInterval{start: 21, end: 100},
+		},
+		{
+			interval:      &reachabilityInterval{start: 1, end: 100},
+			fraction:      0,
+			expectedLeft:  &reachabilityInterval{start: 1, end: 0},
+			expectedRight: &reachabilityInterval{start: 1, end: 100},
+		},
+		{
+			interval:      &reachabilityInterval{start: 1, end: 100},
+			fraction:      1,
+			expectedLeft:  &reachabilityInterval{start: 1, end: 100},
+			expectedRight: &reachabilityInterval{start: 101, end: 100},
```

Comment:
Does it make sense? Do we even want to allow 1 fraction?

Reply:
This is how we mark an "empty" interval. We will replace it with something more explicit in a future ticket.